packages/core/src/internal/effectRunnerMap.js

```diff
@@ -0,0 +1,323 @@
+  makeIterator,
+  noop,
+  remove,
+  shouldComplete,
+} from './utils'
+
+function getIteratorMetaInfo(iterator, fn) {
+  if (iterator.isSagaIterator) {
+    return { name: iterator.meta.name }
+  }
+  return getMetaInfo(fn)
+}
+
+function createTaskIterator({ context, fn, args }) {
+  // catch synchronous failures; see #152 and #441
+  try {
+    const result = fn.apply(context, args)
+
+    // i.e. a generator function returns an iterator
+    if (is.iterator(result)) {
+      return result
+    }
+
+    const next = (value = result) => ({
+      value,
+      done: !is.promise(value),
+    })
+
+    return makeIterator(next)
+  } catch (err) {
+    // do not bubble up synchronous failures for detached forks
+    // instead create a failed task. See #152 and #441
+    return makeIterator(() => {
+      throw err
+    })
+  }
+}
+
+export function runPutEffect(env, { channel, action, resolve }, cb, { resolvePromise }) {
+  /**
+   Schedule the put in case another saga is holding a lock.
+   The put will be executed atomically. ie nested puts will execute after
+   this put has terminated.
+   **/
+  asap(() => {
+    let result
+    try {
+      result = (channel ? channel.put : env.dispatch)(action)
+    } catch (error) {
+      cb(error, true)
+      return
+    }
+
+    if (resolve && is.promise(result)) {
+      resolvePromise(result, cb)
+    } else {
+      cb(result)
+    }
+  })
+  // Put effects are non cancellables
+}
+
+export function runTakeEffect(env, { channel = env.stdChannel, pattern, maybe }, cb) {
+  const takeCb = input => {
+    if (input instanceof Error) {
+      cb(input, true)
+      return
+    }
+    if (isEnd(input) && !maybe) {
+      cb(TERMINATE)
+      return
+    }
+    cb(input)
+  }
+  try {
+    channel.take(takeCb, is.notUndef(pattern) ? matcher(pattern) : null)
+  } catch (err) {
+    cb(err, true)
+    return
+  }
+  cb.cancel = takeCb.cancel
+}
+
+export function runCallEffect(env, { context, fn, args }, cb, { effectId, resolvePromise, resolveIterator }) {
+  // catch synchronous failures; see #152
+  try {
+    const result = fn.apply(context, args)
+
+    if (is.promise(result)) {
+      resolvePromise(result, cb)
+      return
+    }
+
+    if (is.iterator(result)) {
+      resolveIterator(result, effectId, getMetaInfo(fn), cb)
+      return
+    }
+
+    cb(result)
+  } catch (error) {
+    cb(error, true)
+  }
+}
+
+export function runCPSEffect(env, { context, fn, args }, cb) {
+  // CPS (ie node style functions) can define their own cancellation logic
+  // by setting cancel field on the cb
+
+  // catch synchronous failures; see #152
+  try {
+    const cpsCb = (err, res) => (is.undef(err) ? cb(res) : cb(err, true))
+    fn.apply(context, args.concat(cpsCb))
+    if (cpsCb.cancel) {
+      cb.cancel = () => cpsCb.cancel()
+    }
+  } catch (error) {
+    cb(error, true)
+  }
+}
+
+export function runForkEffect(env, { context, fn, args, detached }, cb, { effectId, taskContext, taskQueue }) {
```

Comment:
the same question: is it okay to have a `cb` as not last argument?

Reply:
I think it is just fine. You can think of `cb` as required (every effect runner will use it), and think of the arguments after `cb` as optional (only a few runners will use them), and required argument comes before optional arguments.

Reply:
We may refine the order if we have better alternatives 😄 .